File: 백준/Gold/1958. LCS 3/LCS 3.py
a = [0]+ list(input())
b = [0]+list(input())
c = [0]+list(input())

# i가 a, j가 b, k가 c의 순서가 되기 위해 c가 낮은 차원으로 
dp = [[[0 for _ in range(len(c)+1)] for _ in range(len(b)+1)]
      for _ in range(len(a)+1)]
# 문자열을 순회하며
for i in range(1, len(a)):
    for j in range(1, len(b)):
        for k in range(1, len(c)):
            # 셋다 같은 글자 발견시 +1
            if a[i] == b[j] == c[k]:
                dp[i][j][k] = dp[i-1][j-1][k-1] + 1
            # 아니면 세개 중 큰 값으로 갱신
            else:
                dp[i][j][k] = max(dp[i-1][j][k], dp[i][j-1][k], dp[i][j][k-1])

print(dp[-2][-2][-2])

File: 백준/Gold/1958. LCS 3/LCS 3.java
import java.util.*;

public class Main {
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        // 입력
        String a = "0"+sc.nextLine();
        String b = "0"+sc.nextLine();
        String c = "0"+sc.nextLine();

        int [][][]dp = new int[a.length()][b.length()][c.length()];

        for(int i = 1; i < a.length(); i++){
            for(int j = 1; j < b.length(); j++){
                for(int k = 1; k < c.length(); k++){
                    if (a.charAt(i) == b.charAt(j) && b.charAt(j)== c.charAt(k)){
                        dp[i][j][k] = dp[i-1][j-1][k-1]+1;
                    } else {
                        dp[i][j][k] = Math.max(Math.max(dp[i-1][j][k], dp[i][j-1][k]), dp[i][j][k-1]);

                    }
                }
            }
        }
        System.out.println(dp[a.length()-1][b.length()-1][c.length()-1]);
    }
}
